src/tests/ke/suite/dpc.c

```diff
@@ -0,0 +1,279 @@
+#include <xboxkrnl/xboxkrnl.h>
+#include <string.h>
+
+#include "util/output.h"
+#include "assertions/defines.h"
+
+// TODO: Add list of _LEVEL suffix support into nxdk
+#define DPC_DISPATCH_LEVEL 2 // DISPATCH_LEVEL
+
+typedef struct _dpc_test {
+    BOOLEAN inserted;
+    BOOLEAN removal;
+    BOOL dpc_called;
+    KDPC dpcObject;
+} dpc_test;
+
+static ULONG getDpcRoutineActive()
+{
+    ULONG getter;
+    __asm__ ("movl %%fs:0x58, %0"
+            : "=r"(getter));
+    return getter;
+}
+
+static void setDpcRoutineActive(ULONG setter)
+{
+    __asm__ ("movl %0, %%fs:0x58"
+            :
+            : "a"(setter));
+}
+
+static KDPC dpcObject;
+static BOOL dpc_called;
```

Comment:
This, and all variations of it, must probably be some kind of atomic or marked `volatile`, because it could change anytime, without the compiler knowing about it.

Reply:
How is it needed when interrupt level will call dpc callback once KfLowerIrql is called? See example of test code what to expect:
https://github.com/Cxbx-Reloaded/xbox_kernel_test_suite/blob/570f45d652a487eaa4b5515a2a780c345f0d8046/src/tests/ke/suite/dpc.c#L140-L149
Basically higher IRQL's threads will start if has interrupt requested which give them higher priority to run them first before return to original thread.
- Test 3c is expected to be false even if sleep call been placed there.
- Same thing for test 3d as well.
- So, it is always test 3e will return true no matter what.

One other thing I want to add if you noticed blocks of `GEN_CHECK` is moved to bottom of the test. That is due to getting early triggered unexpected behavior from higher IRQLs lowering down to DISPATCH_LEVEL then to original thread via doing print to screen and/or write to file.

Finally, xbox kernel test suite is only running from one thread. Each test is required to create separate thread and destruction only if specific test needs it. For these tests here, it is not needed. Mainly because xbox kernel will call dpc callback once IRQL is lowering down to original IRQL stored by us from KfRaiseIRQL function.

Reply:
The meaning and function of various "magic values" may be clear to authors, but not to readers. Even authors may forget this and might need to research them when revisiting older code.

One way to avoid that, is to declare symbol names for each value  (perhaps static constant variables or compiler defines), along with an accurate and concise comment to explain their existence with.

Reply:
Each part of the tests already have comments before them. Nothing require anything special from compiler nor constant variables. The main reason these global variables are outside of the function that are being tested is due to they should not be on local stack of current thread as switch to different thread, such as dpc thread, is likely will corrupt the data from their own thread. This is something about readers should at least understand when dealing with multi-threading environment.

Also, it's unnecessary to have such as atomic to make it safe anyway due to KfLowerIRQL call will trigger higher IRQL threads for any interrupt has been requested. Which then will eventually return to owner of the thread.

Perhaps read the comments first for each block of test code before making this conclusion? If there's something not understandable context, please do make review remarks on them instead of here.

Anyway, dpc_ prefix variables are meant to read values from dpc_callback for the test. Which is why they are kept together in same spot. Including the reason of having them as global variables.

Reply:
I'm marking them as invalid due to insufficient reading the whole tests.

Reply:
Btw, it's fine you don't agree with these remarks, but it's inappropriate to conclude reviewers 'not reading the whole code' - perhaps we did and still don't understand why this magic value is here for. Why else would two experienced coders independently conclude a certain review remarks is appropriate?

Reviews are meant to show where a submitted pull request / code changes could be made even better. (Because it's great to see these API's tested! Nothing but praise on that!)